### PR TITLE
DO NOT MERGE YET - Feature/restify docs tests

### PR DESCRIPTION
**Pull request by joelgriffith**
Hey folks, fine project here. This is my first PR, as well as my first TS undertaking, so take that into consideration :)

This introduces a new interface for `restify`, that _almost_ is a copy/pasta of the express integration. I've also taken the liberty of adding some dependencies for this so that things compile and tests run.

Let me know changes you'd like, and whatever else I should take into consideration... and thanks!

- [ ] Merge after https://github.com/apollostack/apollo-server/pull/164

### CODE REVIEW

**Comment by stubailo** [comment]
> that almost is a copy/pasta of the express integration

What's different about it? I've heard of people using `apolloExpress` successfully with Restify.

**Comment by joelgriffith** [comment]
I initially didn't think this would work OOB as the naming suggested it's for `express` only. There's currently an issue here that this would close as well: https://github.com/apollostack/apollo-server/issues/125

This achieves a certain level of integration against restify by documenting it and adding tests. To be fair: I haven't explored the internals of `testSuite(createApp);`, so I'm not sure what it's buying as far as regressions go.

Bascially: instead of folks searching for how to use this with restify, `apollo-server` comes out and says that it supports `restify`.

**Comment by stubailo** [comment]
Sure, I'm just wondering if we could do the same by just saying `export const apolloRestify = apolloExpress` - what's the difference in the implementation?

**Comment by joelgriffith** [comment]
Sure, there's only two differences in implementation at this point

- Reference Types from `@type/restify` as opposed to `@types/express`. If one framework diverges, I'd hope that tests and type annotations would better catch the change as opposed to consumers informing this project.
- Errors messaging: `'POST body missing. Did you forget "server.use(restify.bodyParser())"?'`, which is `restify` specific.

I'm going to continue playing with the implementation locally to see if there's improvements or other considerations to fix, but wanted to get an MVP for this review to get discussion going

**Comment by joelgriffith** [comment]
Apologies on thrashing the pushes, found a few issues when consuming this from an outside project

File: src/integrations/restifyApollo.test.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as restify from 'restify';
```
Comment (joelgriffith) [inline src/integrations/restifyApollo.test.ts]: Doing this makes sure that the interfaces are properly exports as opposed to referencing them directly, so tests a few things in one :)

**Comment by DxCx** [comment]
hi @joelgriffith i think that's a good idea thanks for the contributation!
no worries about messy commits btw, worst case it can be just merged-commit.

however, i'm not sure what's the difference (in implementation) between regular expressApollo to restify one. (from glance look it looks the same)
can you please elaborate?

also, @helfer, can we postpone merging until monorepo is merged?
this will be a new package on top of the monorepo and it will be abit messy to rebase upon it.

**Comment by helfer** [comment]
> can we postpone merging until monorepo is merged?
Sure, I've marked the PR accordingly.

**Comment by joelgriffith** [comment]
I don't mind acting as custodian for this once monorepo is in place. And I definitely echo the sentiment that this is pretty repetitive, and hopefully having clear distinction now makes potential breaking changes easier in the future. 

Lastly, I'll be at the GraphQL summit coming up, and would love say my hellos with this fine team. Thanks!

**Comment by stubailo** [comment]
> Lastly, I'll be at the GraphQL summit coming up, and would love say my hellos with this fine team. Thanks!

Awesome! Excited to see you there!

**Comment by rricard** [comment]
@stubailo I generally agree with @joelgriffith. Express & Restify are very similar but they are effectively different. For now, apolloExpress uses a subset of express features common to restify but let's say in the future, we need to use a super fancy express feature, not available in restify under the same API, then, we're not good!

File: src/integrations/restifyApollo.test.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  it('throws an error if called with more than one argument', function(){
+     expect(() => (<any>apolloRestify)({}, 'x')).to.throw(
+       'Apollo Server expects exactly one argument, got 2');
+  });
+});
+
+describe('integration:Restify', () => {
+  testSuite(createApp);
+});
```
Comment (rricard) [inline src/integrations/restifyApollo.test.ts]: Any help needed to add more tests ? For now, you only ensure the server is not crashing (which is a good start!).

**Comment by DxCx** [comment]
> For now, apolloExpress uses a subset of express features common to restify but let's say in the future, we need to use a super fancy express feature, not available in restify under the same API, then, we're not good!

and yet, IMO, we need to avoid code duplication as much as we can,
at the moment we do not have much difference, and that's why i suggested to lean over each other.
restify will get it's own flavor, which in the future can be changed if needed.

**Comment by rricard** [comment]
@DxCx Well I understand that. If we go that way however, we definitely more tests.

**Comment by DxCx** [comment]
i havn't spoken about testing but implementation itself, IMO, testing is always welcome :)
and ofcourse there should be separated testing for both..

**Comment by rricard** [comment]
@DxCx We completely agree 😉 , I understand that you don't want to avoid duplicating only the implementation! Just saying in this case, we need to ensure that the express implementation stays compatible with restify. And if this implementation were to break, we should then duplicate the code!

**Comment by joelgriffith** [comment]
I can easily add an alias in the export manifest to use the express integration, but just export an additional interface. I think that seems to be the best thing to do at this juncture since there's some concern about duplication.

**Comment by rricard** [comment]
@joelgriffith You will definitely need to export a new interface. I think we found a good compromise with this though! Thanks for contributing anyway, I can't wait to see what you'll do!

**Comment by joelgriffith** [comment]
Sounds good, I'll hold off then atm since I'm fighting the TS type system when trying to alias...

**Comment by joelgriffith** [comment]
Going to re-implement this in monorepo land as I'm having some rought ts conflicts after having migrated.

**Pull request closed without merging**